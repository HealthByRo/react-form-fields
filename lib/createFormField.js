'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

exports.default = createFormField;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _snakeCase = require('lodash/snakeCase');

var _snakeCase2 = _interopRequireDefault(_snakeCase);

var _elements = require('./elements');

var _wrappers = require('./wrappers');

var _Errors = require('./Errors');

var _Errors2 = _interopRequireDefault(_Errors);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

function createFormField(WrappedComponent) {
  return function (_PureComponent) {
    _inherits(FormField, _PureComponent);

    function FormField(props) {
      _classCallCheck(this, FormField);

      var _this = _possibleConstructorReturn(this, (FormField.__proto__ || Object.getPrototypeOf(FormField)).call(this, props));

      if (!props.name || props.name.length < 1) {
        // TODO create file with errors, add MissingParamError
        throw new Error('name param is required for ' + WrappedComponent.name);
      }
      return _this;
    }

    _createClass(FormField, [{
      key: 'render',
      value: function render() {
        var _composeProps = composeProps(this.props),
            labelProps = _composeProps.labelProps,
            inputProps = _composeProps.inputProps,
            errorsProps = _composeProps.errorsProps;

        var showLabel = this.props.label || this.props.labelProps;

        return _react2.default.createElement(
          _wrappers.FieldWrapper,
          null,
          showLabel && _react2.default.createElement(_elements.LabelElement, labelProps),
          _react2.default.createElement(WrappedComponent, inputProps),
          _react2.default.createElement(_Errors2.default, errorsProps)
        );
      }
    }]);

    return FormField;
  }(_react.PureComponent);
}

function composeProps(props) {
  var name = props.name,
      placeholder = props.placeholder,
      label = props.label,
      errors = props.errors,
      inputProps = _objectWithoutProperties(props, ['name', 'placeholder', 'label', 'errors']);

  inputProps.id = inputProps.id || generateInputId(inputProps);
  inputProps.name = name;

  var labelProps = {
    htmlFor: inputProps.id,
    text: label
  };

  var errorsProps = {
    errors: errors
  };

  // don't set placeholder when undefined
  // to avoid placeholder={undefined} in rendered html
  if (placeholder) {
    inputProps.placeholder = placeholder;
  }

  return {
    labelProps: labelProps,
    inputProps: inputProps,
    errorsProps: errorsProps
  };
}

function generateInputId(props) {
  return 'id_' + (0, _snakeCase2.default)(props.name);
}