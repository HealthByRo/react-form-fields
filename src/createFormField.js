/* @flow */

import React, { PureComponent } from 'react';
import snakeCase from 'lodash/snakeCase';
import { LabelElement } from './elements';
import { FieldWrapper } from './wrappers';
import Errors from './Errors';
import type { FieldProps, LabelProps, InputProps, ErrorsProps, Component } from './types';

export default function createFormField(WrappedComponent: Component<*, FieldProps, *>) {
  return class FormField extends PureComponent {
    constructor(props: FieldProps) {
      super(props);

      if (!props.name || props.name.length < 1) {
        // TODO create file with errors, add MissingParamError
        throw new Error(`name param is required for ${WrappedComponent.name}`);
      }
    }

    props: FieldProps;

    render() {
      const {
        labelProps,
        inputProps,
        errorsProps,
      } = composeProps(this.props);
      const showLabel = this.props.label || this.props.labelProps;

      return (
        <FieldWrapper>
          {showLabel &&
            <LabelElement {...labelProps} />
          }
          <WrappedComponent {...inputProps} />
          <Errors {...errorsProps} />
        </FieldWrapper>
      );
    }
  };
}

function composeProps(props: FieldProps): Object {
  const {
    name,
    placeholder,
    label,
  } = props;
  const inputId = props.inputId || generateInputId(props);
  const labelProps: LabelProps = {
    htmlFor: inputId,
    text: label,
  };
  const inputProps: InputProps = {
    name,
    id: inputId,
  };
  const errorsProps: ErrorsProps = {
    errors: props.errors,
  };

  // don't set placeholder when undefined
  // to avoid placeholder={undefined} in rendered html
  if (placeholder) {
    inputProps.placeholder = placeholder;
  }

  return {
    labelProps,
    inputProps,
    errorsProps,
  };
}

function generateInputId(props: FieldProps): string {
  return `id_${snakeCase(props.name)}`;
}
