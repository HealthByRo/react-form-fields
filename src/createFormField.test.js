/* @flow */

import React from 'react';
import renderer from 'react-test-renderer';
import createFormField from './createFormField';
import type { FieldProps } from './types';

const TestField = (props) => (
  <input type="text" {...props} />
);

describe('createFormField(TestField)', () => {
  let tree;
  const FormField = createFormField(TestField);
  const placeholder = 'INPUT_PLACEHOLDER';
  const label = 'INPUT_LABEL';
  const renderWithProps = (props: FieldProps) => {
    tree = renderer.create(
      <FormField {...props} />
    ).toJSON();
  };
  const getPropsWithName = (props?: Object, name?: string = 'INPUT_NAME'): FieldProps => {
    const propsWithName = { ...props };

    if (!propsWithName.name) {
      propsWithName.name = name;
    }

    return propsWithName;
  };

  describe('<TestField />', () => {
    it('should throw error if no name passed', () => {
      // $FlowExpectedError
      expect(() => renderWithProps()).toThrowError('name param is required for TestField');
    });
  });

  describe('<TestField { name } />', () => {
    it('should NOT throw any errors if name is passed', () => {
      expect(() => <TestField name="TEST_NAME" />).not.toThrowError();
    });

    it('should render only wrapped input field with defaults', () => {
      renderWithProps(getPropsWithName());

      expect(tree).toMatchSnapshot();
    });
  });

  describe('<TestField { name, placeholder } />', () => {
    it('should render input with placeholder', () => {
      renderWithProps(getPropsWithName({ placeholder }));

      expect(tree).toMatchSnapshot();
    });
  });

  describe('<TestField { name, placeholder, errors: [SINGLE_ERROR] } />', () => {
    it('should render input with placeholder and single error', () => {
      renderWithProps(getPropsWithName({ placeholder, errors: ['SINGLE_ERROR'] }));

      expect(tree).toMatchSnapshot();
    });
  });

  describe('<TestField { name, placeholder, errors: [FIRST_ERROR, SECOND_ERROR] } />', () => {
    it('should render input with placeholder and two error', () => {
      renderWithProps(getPropsWithName({ placeholder, errors: ['FIRST_ERROR', 'SECOND_ERROR'] }));

      expect(tree).toMatchSnapshot();
    });
  });

  describe('<TestField { name, label } />', () => {
    it('should render input and label', () => {
      renderWithProps(getPropsWithName({ label }));

      expect(tree).toMatchSnapshot();
    });
  });

  describe('<TestField { name, label, errors: [SINGLE_ERROR] } />', () => {
    it('should render input and label and single error', () => {
      renderWithProps(getPropsWithName({ label, errors: ['SINGLE_ERROR'] }));

      expect(tree).toMatchSnapshot();
    });
  });

  describe('<TestField { name, label, errors: [FIRST_ERROR, SECOND_ERROR] } />', () => {
    it('should render input and label and two error', () => {
      renderWithProps(getPropsWithName({ label, errors: ['FIRST_ERROR', 'SECOND_ERROR'] }));

      expect(tree).toMatchSnapshot();
    });
  });

  describe('<TestField { name, label, placeholder } />', () => {
    it('should render input with placeholder and label', () => {
      renderWithProps(getPropsWithName({ label, placeholder }));

      expect(tree).toMatchSnapshot();
    });
  });

  describe('<TestField { name, label, placeholder, errors: [SINGLE_ERROR] } />', () => {
    it('should render input with placeholder, label and single error', () => {
      renderWithProps(getPropsWithName({ label, placeholder, errors: ['SINGLE_ERROR'] }));

      expect(tree).toMatchSnapshot();
    });
  });

  describe('<TestField { name, label, placeholder, errors: [FIRST_ERROR, SECOND_ERROR]  } />', () => {
    it('should render input with placeholder, label and two errors', () => {
      renderWithProps(getPropsWithName({ label, placeholder, errors: ['FIRST_ERROR', 'SECOND_ERROR'] }));

      expect(tree).toMatchSnapshot();
    });
  });
});
