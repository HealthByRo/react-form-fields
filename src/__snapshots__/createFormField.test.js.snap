// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createFormField(TestField) <TestField { name } /> should render only wrapped input field with defaults 1`] = `
<div
  className="form-item"
>
  <input
    inputId="id_input_name"
    name="INPUT_NAME"
    type="text"
  />
</div>
`;

exports[`createFormField(TestField) <TestField { name, label } /> should render input and label 1`] = `
<div
  className="form-item"
>
  <label
    htmlFor="id_input_name"
  >
    INPUT_LABEL
  </label>
  <input
    inputId="id_input_name"
    label="INPUT_LABEL"
    name="INPUT_NAME"
    type="text"
  />
</div>
`;

exports[`createFormField(TestField) <TestField { name, label, errors: [FIRST_ERROR, SECOND_ERROR] } /> should render input and label and two error 1`] = `
<div
  className="form-item"
>
  <label
    htmlFor="id_input_name"
  >
    INPUT_LABEL
  </label>
  <input
    errors={
      Array [
        "FIRST_ERROR",
        "SECOND_ERROR",
      ]
    }
    inputId="id_input_name"
    label="INPUT_LABEL"
    name="INPUT_NAME"
    type="text"
  />
  <div
    className="errors-list"
  >
    <div
      className="msg msg--error"
    >
      FIRST_ERROR
    </div>
    <div
      className="msg msg--error"
    >
      SECOND_ERROR
    </div>
  </div>
</div>
`;

exports[`createFormField(TestField) <TestField { name, label, errors: [SINGLE_ERROR] } /> should render input and label and single error 1`] = `
<div
  className="form-item"
>
  <label
    htmlFor="id_input_name"
  >
    INPUT_LABEL
  </label>
  <input
    errors={
      Array [
        "SINGLE_ERROR",
      ]
    }
    inputId="id_input_name"
    label="INPUT_LABEL"
    name="INPUT_NAME"
    type="text"
  />
  <div
    className="msg msg--error"
  >
    SINGLE_ERROR
  </div>
</div>
`;

exports[`createFormField(TestField) <TestField { name, label, placeholder } /> should render input with placeholder and label 1`] = `
<div
  className="form-item"
>
  <label
    htmlFor="id_input_name"
  >
    INPUT_LABEL
  </label>
  <input
    inputId="id_input_name"
    label="INPUT_LABEL"
    name="INPUT_NAME"
    placeholder="INPUT_PLACEHOLDER"
    type="text"
  />
</div>
`;

exports[`createFormField(TestField) <TestField { name, label, placeholder, errors: [FIRST_ERROR, SECOND_ERROR]  } /> should render input with placeholder, label and two errors 1`] = `
<div
  className="form-item"
>
  <label
    htmlFor="id_input_name"
  >
    INPUT_LABEL
  </label>
  <input
    errors={
      Array [
        "FIRST_ERROR",
        "SECOND_ERROR",
      ]
    }
    inputId="id_input_name"
    label="INPUT_LABEL"
    name="INPUT_NAME"
    placeholder="INPUT_PLACEHOLDER"
    type="text"
  />
  <div
    className="errors-list"
  >
    <div
      className="msg msg--error"
    >
      FIRST_ERROR
    </div>
    <div
      className="msg msg--error"
    >
      SECOND_ERROR
    </div>
  </div>
</div>
`;

exports[`createFormField(TestField) <TestField { name, label, placeholder, errors: [SINGLE_ERROR] } /> should render input with placeholder, label and single error 1`] = `
<div
  className="form-item"
>
  <label
    htmlFor="id_input_name"
  >
    INPUT_LABEL
  </label>
  <input
    errors={
      Array [
        "SINGLE_ERROR",
      ]
    }
    inputId="id_input_name"
    label="INPUT_LABEL"
    name="INPUT_NAME"
    placeholder="INPUT_PLACEHOLDER"
    type="text"
  />
  <div
    className="msg msg--error"
  >
    SINGLE_ERROR
  </div>
</div>
`;

exports[`createFormField(TestField) <TestField { name, placeholder } /> should render input with placeholder 1`] = `
<div
  className="form-item"
>
  <input
    inputId="id_input_name"
    name="INPUT_NAME"
    placeholder="INPUT_PLACEHOLDER"
    type="text"
  />
</div>
`;

exports[`createFormField(TestField) <TestField { name, placeholder, errors: [FIRST_ERROR, SECOND_ERROR] } /> should render input with placeholder and two error 1`] = `
<div
  className="form-item"
>
  <input
    errors={
      Array [
        "FIRST_ERROR",
        "SECOND_ERROR",
      ]
    }
    inputId="id_input_name"
    name="INPUT_NAME"
    placeholder="INPUT_PLACEHOLDER"
    type="text"
  />
  <div
    className="errors-list"
  >
    <div
      className="msg msg--error"
    >
      FIRST_ERROR
    </div>
    <div
      className="msg msg--error"
    >
      SECOND_ERROR
    </div>
  </div>
</div>
`;

exports[`createFormField(TestField) <TestField { name, placeholder, errors: [SINGLE_ERROR] } /> should render input with placeholder and single error 1`] = `
<div
  className="form-item"
>
  <input
    errors={
      Array [
        "SINGLE_ERROR",
      ]
    }
    inputId="id_input_name"
    name="INPUT_NAME"
    placeholder="INPUT_PLACEHOLDER"
    type="text"
  />
  <div
    className="msg msg--error"
  >
    SINGLE_ERROR
  </div>
</div>
`;
